csrc/cpu/neighbor_sample_cpu.cpp

```diff
@@ -199,31 +224,58 @@ hetero_sample(const vector<node_t> &node_types,
           continue;
 
         if ((num_samples < 0) || (!replace && (num_samples >= col_count))) {
+          // select all neighbors
           for (int64_t offset = col_start; offset < col_end; offset++) {
             const int64_t &v = row_data[offset];
+            bool time_constraint = true; // whether src -> dst obeys the time constraint
+            if (is_temporal) {
+              const auto *src_time = node_time_dict.at(src_node_type).data_ptr<int64_t>();
+              if (dst_time < src_time[v])
+                time_constraint = false;
+            }
+            if (!time_constraint)
+              continue;
             const auto res = to_local_src_node.insert({v, src_samples.size()});
-            if (res.second)
+            if (res.second) {
               src_samples.push_back(v);
+              if (is_temporal)
+                src_root_time.push_back(dst_time);
+            }
             if (directed) {
               cols.push_back(i);
               rows.push_back(res.first->second);
               edges.push_back(offset);
             }
           }
         } else if (replace) {
-          for (int64_t j = 0; j < num_samples; j++) {
+          // sample with replacement
+          int64_t num_neighbors = 0;
+          while (num_neighbors < num_samples) {
             const int64_t offset = col_start + rand() % col_count;
             const int64_t &v = row_data[offset];
+            bool time_constraint = true; // whether src -> dst obeys the time constraint
+            if (is_temporal) {
+              const auto *src_time = node_time_dict.at(src_node_type).data_ptr<int64_t>();
+              if (dst_time < src_time[v])
+                time_constraint = false;
+            }
+            if (!time_constraint)
+              continue;
             const auto res = to_local_src_node.insert({v, src_samples.size()});
-            if (res.second)
+            if (res.second) {
               src_samples.push_back(v);
+              if (is_temporal)
+                src_root_time.push_back(dst_time);
+            }
             if (directed) {
               cols.push_back(i);
               rows.push_back(res.first->second);
               edges.push_back(offset);
             }
+            num_neighbors += 1;
           }
         } else {
+          // sample without replacement
           unordered_set<int64_t> rnd_indices;
           for (int64_t j = col_count - num_samples; j < col_count; j++) {
```

Comment:
If selected nodes < `num_samples`, do we need to look at more nodes?

Reply:
For sampling without replacement, if selected nodes < num_samples, we will just return the maximum number of neighbors obeying time constraint. For the sampling with replacement, we can of course resample.
What do you think? Do you have any suggestion?

Reply:
There are two approaches:

- Use original sampling logic and use timestamp to filter out invalid ones from the result
- Filter out invalid neighbors and do the original sampling

I'm not sure which is better from the accuracy point of view. You seem to follow the first one, and it's good to me. What do you think about the second one?

Reply:
Yeah i think both are equivalent in terms of sampling outcomes, since the first option is a simplified rejection sampling.
And efficiency-wise the first option requires one for loop and the second one requires 2 for loops (but each iteration does slightly less things). So I think the difference will be pretty small as well.